File: android/app/src/main/java/com/xirigo/ecommerce/core/domain/error/AppError.kt
package com.xirigo.ecommerce.core.domain.error

import com.xirigo.ecommerce.R

sealed class AppError : Exception() {
    data class Network(override val message: String = "Network error") : AppError()
    data class Server(val code: Int, override val message: String) : AppError()
    data class NotFound(override val message: String = "Not found") : AppError()
    data class Unauthorized(override val message: String = "Unauthorized") : AppError()
    data class Unknown(override val message: String = "Unknown error") : AppError()
}

fun Throwable.toUserMessageResId(): Int = when (this) {
    is AppError.Network -> R.string.common_error_network
    is AppError.Server -> R.string.common_error_server
    is AppError.Unauthorized -> R.string.common_error_unauthorized
    is AppError.NotFound -> R.string.common_error_not_found
    else -> R.string.common_error_unknown
}

File: ios/XiriGoEcommerce/Core/Domain/Error/AppError.swift
import Foundation

// MARK: - AppError

enum AppError: Error, Equatable {
    case network(message: String = "Network error")
    case server(code: Int, message: String)
    case notFound(message: String = "Not found")
    case unauthorized(message: String = "Unauthorized")
    case unknown(message: String = "Unknown error")
}

// MARK: - User Message Mapping

extension Error {
    /// Returns localized user-facing message from String Catalog
    var toUserMessage: String {
        guard let appError = self as? AppError else {
            return String(localized: "common_error_unknown")
        }
        switch appError {
            case .network:
                return String(localized: "common_error_network")

            case .server:
                return String(localized: "common_error_server")

            case .unauthorized:
                return String(localized: "common_error_unauthorized")

            case .notFound:
                return String(localized: "common_error_not_found")

            case .unknown:
                return String(localized: "common_error_unknown")
        }
    }
}
